tuned/plugins/plugin_amdgpu.py

```diff
@@ -0,0 +1,124 @@
+cmd = commands()
+
+UPOWER_DBUS_NAME = "org.freedesktop.UPower"
+UPOWER_DBUS_PATH = "/org/freedesktop/UPower"
+UPOWER_DBUS_INTERFACE = "org.freedesktop.UPower"
+
+
+class AmdgpuPlugin(base.Plugin):
+    """
+    `amdgpu`::
+
+    Configures panel power savings on laptops with amdgpu driven eDP panels.
+    This accepts a value range from 0 to 4, where 4 is the highest power savings
+    but will trade off color accuracy.
+
+    Settings will only be applied when the system is on battery.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(AmdgpuPlugin, self).__init__(*args, **kwargs)
+        self.proxy = None
+
+    def upower_changed(self, interface, changed, invalidated):
+        properties = dbus.Interface(self.proxy, "org.freedesktop.DBus.Properties")
+        self._on_battery = bool(properties.Get(UPOWER_DBUS_INTERFACE, "OnBattery"))
+        log.debug("🔋: %s, 🎯: %s" % (self._on_battery, self.target_value))
+        for device in self._assigned_devices:
+            self.apply_target(device)
+
+    def setup_battery_signaling(self):
+        bus = dbus.SystemBus()
+        self.proxy = bus.get_object(UPOWER_DBUS_NAME, UPOWER_DBUS_PATH)
+        self.proxy.connect_to_signal("PropertiesChanged", self.upower_changed)
+        self.upower_changed(None, None, None)
+
+    def _init_devices(self):
+        self._assigned_devices = set()
+        self._free_devices = set()
+
+        for device in (
+            self._hardware_inventory.get_devices("drm")
+            .match_sys_name("card*eDP*")
+            .match_sys_name("amdgpu")
+            .match_sys_name("panel_power_savings")
+        ):
+            self._free_devices.add(device.sys_name)
+        self._devices_supported = len(self._free_devices) > 0
+
+    def _get_device_objects(self, devices):
+        return [self._hardware_inventory.get_device("drm", x) for x in devices]
+
+    @classmethod
+    def _get_config_options(cls):
+        return {"panel_power_savings": None}
+
+    def _instance_init(self, instance):
+        instance._has_static_tuning = True
+        instance._has_dynamic_tuning = False
+
+    def _amdgpu_files(self, device):
+        return {
+            "panel_power_savings": "/sys/class/drm/%s/amdgpu/panel_power_savings" % device
+        }
+
+    def _instance_cleanup(self, instance):
+        pass
+
+    def apply_target(self, device):
+        """Apply the target value to the panel_power_savings file if it doesn't already have it"""
+
+        # if we don't have a proxy, we can't tell if we're on battery
+        if not self.proxy:
+            self.setup_battery_signaling()
+            return None
+
+        # determine value to use (only apply if on battery, otherwise set to 0)
+        if self._on_battery:
+            target = self.target_value
+        else:
+            target = 0
+
+        # make sure the value is different (avoids unnecessary kernel modeset)
+        current = int(self._get_panel_power_savings(device))
```

Comment:
This fails when the file `/sys/class/drm/card1-eDP-1/amdgpu/panel_power_savings` does not exist.

Reply:
But doesn't the plugin fail to add any devices?

Reply:
Oh, right, that should suffice when implemented correctly. Currently, you do:
```python
self._hardware_inventory.get_devices("drm")
            .match_sys_name("card*eDP*")
            .match_sys_name("amdgpu")
            .match_sys_name("panel_power_savings")
```
Applying `match_sys_name` multiple times yields all devices that match at least one of the names (i.e., it's a union, not an intersection). In my case, `card1-eDP-1` gets matched and added to plugin devices.

Reply:
Ah! Got it.  I've pushed a change to only match the eDP and add the device if the sysfs file is found for it.